core/include/vecmem/utils/impl/copy.ipp

```diff
@@ -395,6 +398,138 @@ copy::event_type copy::set_sizes(
+                edm::type::details::is_same_nc<VARTYPES1, VARTYPES2>...>,
+        "Can only use compatible types in the copy");
+
+    // First, handle the simple case, when both views have a contiguous memory
+    // layout.
+    if ((from_view.payload().ptr() != nullptr) &&
+        (to_view.payload().ptr() != nullptr) &&
+        (from_view.payload().size() == to_view.payload().size())) {
+
+        // If the "common size" is zero, we're done.
+        if (from_view.payload().size() == 0) {
+            return vecmem::copy::create_event();
+        }
+
+        // Let the user know what's happening.
+        VECMEM_DEBUG_MSG(2, "Performing simple SoA copy of %u bytes",
+                         from_view.payload().size());
+
+        // Copy the payload with a single copy operation.
+        operator()(from_view.payload(), to_view.payload(), cptype);
+
+        // If the target view is resizable, set its size.
+        if (to_view.size().ptr() != nullptr) {
+            // If the source is also resizable, the situation should be simple.
+            if (from_view.size().ptr() != nullptr) {
+                // Check that the sizes are the same.
+                if (from_view.size().size() != to_view.size().size()) {
+                    throw std::runtime_error(
+                        "Copying between inconsistent containers");
+                }
+                // Perform a dumb copy.
+                operator()(from_view.size(), to_view.size(), cptype);
+            } else {
+                // If not, then copy the size(s) recursively.
+                copy_sizes_impl(from_view, to_view, cptype,
+                                std::index_sequence_for<VARTYPES1...>{});
+            }
+        }
+
+        // Create a synhronization event.
+        return create_event();
+    }
+
+    // If not, then do an un-optimized copy, variable-by-variable.
+    copy_payload_impl(from_view, to_view, cptype,
+                      std::index_sequence_for<VARTYPES1...>{});
+
+    // Return a new event.
+    return create_event();
+}
+
+template <typename... VARTYPES1, typename... VARTYPES2>
+copy::event_type copy::operator()(
+    const edm::view<edm::schema<VARTYPES1...>>& from_view,
+    edm::host<edm::schema<VARTYPES2...>>& to_vec,
+    type::copy_type cptype) const {
+
+    // The input and output types are allowed to be different, but only by
+    // const-ness.
+    static_assert(sizeof...(VARTYPES1) == sizeof...(VARTYPES2),
+                  "Can only use compatible types in the copy");
+    static_assert(std::is_same<edm::schema<VARTYPES1...>,
+                               edm::schema<VARTYPES2...>>::value ||
+                      details::is_same_nc<edm::schema<VARTYPES1...>,
+                                          edm::schema<VARTYPES2...>>::value,
+                  "Can only use compatible types in the copy");
+
+    // Resize the output object to the correct size(s).
+    resize_impl(from_view, to_vec, cptype,
```

Comment:
Ah I see, in this case it does. So view -> view assumes it's already correctly sized, and view->vector will resize, then call view -> view ?

Reply:
Exactly. That's the logic that the code has been using for 1D and jagged vectors as well. :wink: So that copying is only implemented between two views. Regardless of what exact type of object is holding the memory.